src/GLSL.hs

```diff
@@ -0,0 +1,75 @@
+module GLSL where
+import Syntax (OpaqueType)
+
+data GSize = Two
+           | Three
+           | Four
+        deriving (Eq, Ord)
+
+data GType = Void
+           | GBool
+           | GInt
+           | GUInt
+           | GFloat
+           | GDouble
+           | GVec GSize GType
+           | GMat GSize GSize
+
+data GOp = GArr
+         | GField
+         | GMult
+         | GMod
+         | GDiv
+         | GAdd
+         | GSub
+         | GRShift
+         | GLShift
+         | GLt
+         | GLte
+         | GGt
+         | GGte
+         | GEq
+         | GNEq
+         | GAnd
+         | GOr
+         | GXOr
+         | GBAnd
+         | GBOr
+         | GBXOr
+
+data GUOp = GFLip
+          | GNot
+          | GNeg
+          | GPos
+
+data GConst = GR String
+            | GI Int
+            | GB Bool
+            | GF Float
+            | GD Double
+            | GV [GConst]
+            | GM [[GConst]]
+
+data GExpr = GCall String [GExpr]
+           | GParen GExpr
+           | GAssign String GExpr
+           | Imm GConst
+           | GUn GUOp GExpr
+           | GBin GOp GExpr GExpr
+           | GTern GExpr GExpr GExpr
```

Comment:
I think working with arrays could qualify as binary ops, but the lefthand side always needs to be a name for that to work. A single case for indexing a name would work, with the following expr resolving to either a var or an int.

Reply:
Yes, if you look at the `GOp` list I had considered that! Though you make a good point that it must always be a name. I had given some thought to make the `GOp` list a GADT so as to enforce this at the Haskell type level, but then I remembered the issue with references.

Reply:
I was playing around and found you could alter type annotations through a function like this
```haskell
trans :: Expr a -> b -> Expr b
trans (Ref e) _ = Ref e
```
Reminds me of writing Idris!